clang/tools/checked-c-convert/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -1,127 +1,558 @@
     if (VarDecl *VD = dyn_cast<VarDecl>(D)) {
       Expr *InitE = VD->getInit();
-      Expr *sizeArg;
-      if (isAllocatorCall(InitE, &sizeArg)) {
-        if (Info.isIdentifiedArrayVar(D))
-          Info.addAllocationBasedSizeExpr(D, sizeArg);
-        else
-          dumpNotArrayIdentifiedVariable(D, InitE, llvm::dbgs());
+      if (needArrayBounds(VD, Info, Context) &&
+          InitE && isAllocatorCall(InitE)) {
+        ArrBInfo.addBoundsInformation(VD,
+                                           getAllocatedSizeExpr(InitE,
+                                                                Context,
+                                                                Info));
       }
     }
 
   return true;
 }
 
-// check if the provided expression is a call to known memory allocators.
-// if yes, return true along with the argument used as the size assigned
-// to the second parameter i.e., sizeArgument
-bool LocalVarABVisitor::isAllocatorCall(Expr *currExpr, Expr **sizeArgument) {
-  if (currExpr != nullptr) {
-    currExpr = removeAuxiliaryCasts(currExpr);
-    // check if this is a call expression to a named function.
-    if (CallExpr *CA = dyn_cast<CallExpr>(currExpr))
-      if (CA->getCalleeDecl() != nullptr) {
-        FunctionDecl *calleeDecl = dyn_cast<FunctionDecl>(CA->getCalleeDecl());
-        if (calleeDecl) {
-          StringRef funcName = calleeDecl->getName();
-          // check if the called function is a known allocator?
-          if (LocalVarABVisitor::AllocatorFunctionNames.find(funcName) !=
-              LocalVarABVisitor::AllocatorFunctionNames.end()) {
-            if (sizeArgument != nullptr)
-              *sizeArgument = CA->getArg(0);
-            return true;
-          }
-        }
-      }
-  }
-  return false;
-}
+bool LocalVarABVisitor::VisitSwitchStmt(SwitchStmt *S) {
+  VarDecl *CondVar = S->getConditionVariable();
 
-// check if expression is a simple local variable
-// i.e., ptr = .
-// if yes, return the referenced local variable as the return
-// value of the argument.
-bool LocalVarABVisitor::isExpressionSimpleLocalVar(Expr *toCheck, Decl **targetDecl) {
-  if (DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(toCheck))
-    if (DeclaratorDecl *FD = dyn_cast<DeclaratorDecl>(DRE->getDecl()))
-      if (Decl *V = dyn_cast<Decl>(FD)) {
-        *targetDecl = V;
-        return true;
-      }
-  return false;
+  // If this is a parameter declaration? Then this parameter cannot be length.
+  if (CondVar != nullptr)
+    if (ParmVarDecl *PD = dyn_cast<ParmVarDecl>(CondVar))
+      NonLengthParameters.insert(PD);
 }
 
-void LocalVarABVisitor::dumpNotArrayIdentifiedVariable(Decl *LHS, Expr *RHS, raw_ostream &O) {
-#ifdef DEBUG
-  O << "Not identified as a array variable.\n RHS:";
-  RHS->dump(O);
-  O << "\n LHS:";
-  LHS->dump(O);
-  O << "\n";
-#endif
+// Check if the provided parameter cannot be a length of an array.
+bool LocalVarABVisitor::isNonLengthParameter(ParmVarDecl* PVD) {
```

Comment:
* aligns with the variable name instead of the type name. Use `T *x` instead of `T* x`.